### PR TITLE
Support macro provide

### LINKED ISSUE #346: Support methods requiring implicit ev (`NeedsEnv`, `CanFail`)
In ZIO 2.0 those methods (e.g. `injectSome`) are going through a macro, which means they no longer themselves take implicit parameters. Which makes it statically impossible to prove in order to show the e.g. `InfallibleEffectRecoveryInspection`

One possibility is to create a synthetic member injector and "inject" ZIO with those signatures as if they were still present on the ZIO object. Not sure how this could be resolved, though.

### PR DESCRIPTION
**Pull request by myazinn**
Partially closes https://github.com/zio/zio-intellij/issues/346
~~Copypasted~~ carefully ported version of [LayerBuilder in ZIO](https://github.com/zio/zio/blob/series/2.x/core/shared/src/main/scala/zio/internal/macros/LayerBuilder.scala).
For now, it supports only `.provide` method. Addition of `.provideSome` should be easy, but I decided to do it in a separate MR.
There are some tricky places (I'll highlight them in comments), but other then that it should work fine

This implementation is quite straightforward and provides only hints and messages without quick fixes. Perhaps someone can do it another day if needed.
image
image
image

### CODE REVIEW

File: src/main/scala/zio/intellij/utils/TypeCheckUtils.scala [hunk 1]
```diff
@@ -8,7 +8,7 @@ import zio.intellij.utils.types.{ZLayerTypes, ZioTypes}
 
 object TypeCheckUtils {
 
-  val zioTypes        = ZioTypes.values.map(_.fqName)
+  val zioTypes        = ZioTypes.values.map(_.fqName) :+ "zio.ZIOVersionSpecific"
```
Comment (myazinn) [inline src/main/scala/zio/intellij/utils/TypeCheckUtils.scala:11]: `.provide` method is defined in this trait and without this change pattern matching wasn't working :(

File: src/main/scala/zio/intellij/inspections/macros/ProvideMacroZIO2Inspection.scala [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+  // dirty hack to make it work
+  // ScType and ScExpression don't have equals / hashCode methods, which makes it difficult to use it with the algorithm
+  // besides, original ZIO algorithm uses string comparison too
+  final class ZType(val value: ScType)(implicit context: TypePresentationContext) {
+    def conforms(that: ZType): Boolean = this.value.conforms(that.value)
+
+    override def equals(other: Any): Boolean = other match {
+      case that: ZType => this.asStr == that.asStr
+      case _           => false
+    }
```
Comment (myazinn) [inline src/main/scala/zio/intellij/inspections/macros/ProvideMacroZIO2Inspection.scala]: Unstable way of comparing types. Could find another way of doing it :(

**Comment by hmemcpy** [comment]
I always wanted to do this but was too lazy :D So thank you so much!
Could you also hack in support for zio-magic's `.inject*` for this? Should work the same, I believe.

**Comment by myazinn** [comment]
Yes, I think it'll be easy to do

**Comment by hmemcpy** [comment]
This looks good even with the small remarks! You can either merge it as-is, or after adding the `.inject` support!

**Comment by myazinn** [comment]
@hmemcpy thanks
I think I'll finish my work on `.inject` today, so maybe it's better to merge them in one MR

**Comment by myazinn** [comment]
@hmemcpy I've added a new commit with `.inject` support

**Comment by hmemcpy** [comment]
Amazing! Thank you so much!

**Pull request merged**